File: src/SIStorage.Service.Client/FacetsApi.cs
using SIStorage.Service.Contract;
using SIStorage.Service.Contract.Models;
using System.Net.Http.Json;

namespace SIStorage.Service.Client;

internal sealed class FacetsApi : IFacetsApi
{
    private readonly HttpClient _client;

    /// <summary>
    /// Initializes a new instance of <see cref="FacetsApi" /> class.
    /// </summary>
    /// <param name="client">HTTP client to use.</param>
    public FacetsApi(HttpClient client) => _client = client;

    public async Task<Publisher[]> GetPublishersAsync(int? languageId = null, CancellationToken cancellationToken = default) =>
        (await _client.GetFromJsonAsync<Publisher[]>($"facets/publishers{CreateLanguageParam(languageId)}", cancellationToken)) ?? Array.Empty<Publisher>();

    public async Task<Author[]> GetAuthorsAsync(int? languageId = null, CancellationToken cancellationToken = default) =>
        (await _client.GetFromJsonAsync<Author[]>($"facets/authors{CreateLanguageParam(languageId)}", cancellationToken)) ?? Array.Empty<Author>();

    public async Task<Tag[]> GetTagsAsync(int? languageId = null, CancellationToken cancellationToken = default) =>
        (await _client.GetFromJsonAsync<Tag[]>($"facets/tags{CreateLanguageParam(languageId)}", cancellationToken)) ?? Array.Empty<Tag>();

    public async Task<Restriction[]> GetRestrictionsAsync(CancellationToken cancellationToken = default) =>
        (await _client.GetFromJsonAsync<Restriction[]>("facets/restrictions", cancellationToken)) ?? Array.Empty<Restriction>();

    public async Task<Language[]> GetLanguagesAsync(CancellationToken cancellationToken = default) =>
        (await _client.GetFromJsonAsync<Language[]>("facets/languages", cancellationToken)) ?? Array.Empty<Language>();

    private static string CreateLanguageParam(int? languageId) => languageId.HasValue ? $"?languageId={languageId.Value}" : "";
}

File: web/src/FacetsApi.ts
import { getAsync } from './helpers';
import Author from './models/Author';
import Language from './models/Language';
import Publisher from './models/Publisher';
import Restriction from './models/Restriction';
import Tag from './models/Tag';

/** Provides API for working with facets. */
export default class FacetsApi {
	/**
	 * Initializes a new instance of FacetsApi class.
	 * @param baseUri Base uri to send requests to.
	 */
	constructor(private baseUri: string) {}

	/**
	 * Gets well-known package authors.
	 * @param languageId Language identifier.
	 */
	async getAuthorsAsync(languageId?: number) {
		return getAsync<Author[]>(`${this.baseUri}authors${this.createLanguageParameter(languageId)}`);
	}

	/**
	 * Gets well-known languages.
	 */
	async getLanguagesAsync() {
		return getAsync<Language[]>(`${this.baseUri}languages`);
	}

	/**
	 * Gets well-known package tags.
	 * @param languageId Language identifier.
	 */
	async getTagsAsync(languageId?: number) {
		return getAsync<Tag[]>(`${this.baseUri}tags${this.createLanguageParameter(languageId)}`);
	}

	/**
	 * Gets well-known package publishers.
	 * @param languageId Language identifier.
	 */
	async getPublishersAsync(languageId?: number) {
		return getAsync<Publisher[]>(`${this.baseUri}publishers${this.createLanguageParameter(languageId)}`);
	}

	/**
	 * Gets well-known package restrictions.
	 */
	async getRestrictionsAsync() {
		return getAsync<Restriction[]>(`${this.baseUri}restrictions`);
	}

	private createLanguageParameter(languageId?: number): string {
		return languageId === undefined ? '' : `?languageId=${languageId}`;
	}
}